### PR TITLE
global_cycle - Support NoahMP and fractional grid

### LINKED ISSUE #549: global_cycle fractional grid support
The coupled model is now using a fractional grid, but the data assimilation still needs to use global_cycle in order to bring in the land DA analysis increments. Global cycle needs to be updated to be consistent with the code in https://github.com/NCAR/ccpp-physics/pull/695

### PR DESCRIPTION
**Pull request by GeorgeGayno-NOAA**
## DESCRIPTION OF CHANGES: 
Updates to support NoahMP and fractional grids.

## TESTS CONDUCTED: 

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Hercules, Jet, Hera and WCOSS2). Done using 8f1d560.
- [x] Compile branch on Hera using GNU. Done using 8f1d560.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done using 8f1d560.
- [x] Run unit tests locally on any Tier 1 machine. Done on Hera using 8f1d560.
- [x] Run relevant consistency tests locally on all Tier 1 machine. Done using 8f1d560. One global_cycle test failed. See below for details.
- [x] Check Doxygen compile on any Tier 1 machine. Done on Hera using 8f1d560.

Test 2 failed on all machines. The data values in the baseline files and the branch files were the same. However, two records had different NetCDF 'checksum' attributes. This is not a problem.

## DEPENDENCIES:
None.

## DOCUMENTATION:
N/A

## ISSUE: 
Fixes #549.

### CODE REVIEW

**Comment by aerorahul** [comment]
Started testing in global-workflow through CI in https://github.com/NOAA-EMC/global-workflow/pull/1818

**Comment by GeorgeGayno-NOAA** [comment]
@ClaraDraper-NOAA - This update to `global_cycle` supports fractional grids and NoahMP. I just merged your updates from #834 to my branch. Can you please check if your components will work with fractional grids?

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA - This update to `global_cycle` supports fractional grids and NoahMP. I just merged your updates from #834 to my branch. Can you please check if your components will work with fractional grids?

@GeorgeGayno-NOAA Without looking at the code, we will need to make some changes to my snow depth  incrementing code, but I think the soil moisture should be OK. I'll run some tests to make sure. Do you have an example set-up for fractional grids that I could use as a test (I'm mostly just after the restarts)?

**Comment by ClaraDraper-NOAA** [comment]
Also, this is great that we'll be able to use fractional grids finally!

File: sorc/global_cycle.fd/cycle.f90 [hunk 1]
```diff
@@ -444,6 +452,10 @@ SUBROUTINE SFCDRV(LUGB, IDIM,JDIM,LENSFC,LSOIL,DELTSFC,  &
  ENDIF
 
 IF (DO_LNDINC) THEN
+   IF (FRAC_GRID) THEN
```
Comment (GeorgeGayno-NOAA) [inline sorc/global_cycle.fd/cycle.f90:457]: @ClaraDraper-NOAA - looks like I stop processing when DO_LNDINC is chosen with fractional grids. Do you want to fix this now or in a subsequent issue?

Comment (ClaraDraper-NOAA) [reply sorc/global_cycle.fd/cycle.f90:457]: Let's fix it in a separate issue.

**Comment by GeorgeGayno-NOAA** [comment]
> > @ClaraDraper-NOAA - This update to `global_cycle` supports fractional grids and NoahMP. I just merged your updates from #834 to my branch. Can you please check if your components will work with fractional grids?
> 
> @GeorgeGayno-NOAA Without looking at the code, we will need to make some changes to my snow depth incrementing code, but I think the soil moisture should be OK. I'll run some tests to make sure. Do you have an example set-up for fractional grids that I could use as a test (I'm mostly just after the restarts)?

I am using these restarts (on Hera): `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/global_cycle_fraction/input_data.fractional/cpld_bmark_p8_2`

And this test script: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/global_cycle_fraction/run.frac.sh`

**Comment by ClaraDraper-NOAA** [comment]
> > > @ClaraDraper-NOAA - This update to `global_cycle` supports fractional grids and NoahMP. I just merged your updates from #834 to my branch. Can you please check if your components will work with fractional grids?
> > 
> > 
> > @GeorgeGayno-NOAA Without looking at the code, we will need to make some changes to my snow depth incrementing code, but I think the soil moisture should be OK. I'll run some tests to make sure. Do you have an example set-up for fractional grids that I could use as a test (I'm mostly just after the restarts)?
> 
> I am using these restarts (on Hera): `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/global_cycle_fraction/input_data.fractional/cpld_bmark_p8_2`
> 
> And this test script: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/global_cycle_fraction/run.frac.sh`

Thanks. I'll copy those now, but won't get to this until at least next week.

File: sorc/global_cycle.fd/cycle.f90 (hunk 1 above)
Comment (GeorgeGayno-NOAA) [reply sorc/global_cycle.fd/cycle.f90:457]: Sounds good.

**Comment by CatherineThomas-NOAA** [comment]
Hi @GeorgeGayno-NOAA @ClaraDraper-NOAA, I was wondering if either of you have any updates on this?  We're pretty eager to get this update back into the global-workflow for our cycled testing.  Is there any evaluation that I can help with?

**Comment by GeorgeGayno-NOAA** [comment]
> Hi @GeorgeGayno-NOAA @ClaraDraper-NOAA, I was wondering if either of you have any updates on this? We're pretty eager to get this update back into the global-workflow for our cycled testing. Is there any evaluation that I can help with?

Do you plan to use @ClaraDraper-NOAA soil moisture and snow update options?

**Comment by CatherineThomas-NOAA** [comment]
@GeorgeGayno-NOAA  We definitely plan to use the snow DA for v17.  We would like to use the soil DA as well but that's less certain at the moment.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA We definitely plan to use the snow DA for v17. We would like to use the soil DA as well but that's less certain at the moment.

If you want to start testing, but with the snow and soil DA turned off, you can use what is currently in the branch. Otherwise, I will need to consult with @ClaraDraper-NOAA to ensure the snow DA part is working.

**Comment by ClaraDraper-NOAA** [comment]
@CatherineThomas-NOAA @GeorgeGayno-NOAA 

I was on travel last week, but will get to this this week. A few details: 

- I think it's fine for you to start you testing without the soil analysis, since this isn't included in any of your current systems. There is also one more set of PRs I need to push to the GSI for the soil analysis (including name changes of soil variable from Noah to Noah-MP). I can also try to start on this this week.
- which snow DA are your trying to test? The new one (mine, using JEDI), or the current one (handled through global_cycle)? 
- The new snow DA does not currently use UFS_UTILS. UFS_UTILS couldn't handle fractional grids at the time, so I wrote my own program. Jiarui and Cory have merged my program into the global_workflow, so there's no need for you to use UFS_UTILS for that. 
-we haven't yet made a decision as to whether to move all the land DA incrementing code into UFS_UTILS, or keep it separate (starting from my program). I lean towards the latter - in which case we'd only keep using UFS_UTILS/add new code into UFS_UTILS when we're using the GSI (as we need the re-gridding code from the GSI grid to the model grid that is  in UFS_UTILS). Once the atmos DA switches to JEDI, we'd stop using UFS_UTILS to add the increments.

**Comment by ClaraDraper-NOAA** [comment]
@CatherineThomas-NOAA one more question: which land model are you using?

**Comment by CatherineThomas-NOAA** [comment]
@ClaraDraper-NOAA  Thanks for chiming in.  For an intermediate v17 target without the soil DA, we will be using Noah-MP for the land model and the new JEDI snow DA.  However, our immediate goal is to be able to use this updated global_cycle in the global workflow with an ATM-only model, new physics (including fractional grid and Noah-MP), and GSI ATM DA.  Essentially being able to use global workflow develop to cycle something that looks somewhat like current operations but with updated physics.

Concerning your last point about where the keep the land incrementing code, there are ongoing discussions about where to keep many of the DA-adjacent scripts (for example [g-w #1839](https://github.com/NOAA-EMC/global-workflow/issues/1839)), so I'd imagine that the land incrementing code would fall under similar standards.  If you would like to be part of those discussions, let me or other key DA folks (Cory/Daryl/etc) know.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Thanks for chiming in. For an intermediate v17 target without the soil DA, we will be using Noah-MP for the land model and the new JEDI snow DA. However, our immediate goal is to be able to use this updated global_cycle in the global workflow with an ATM-only model, new physics (including fractional grid and Noah-MP), and GSI ATM DA. Essentially being able to use global workflow develop to cycle something that looks somewhat like current operations but with updated physics.
> 
> Concerning your last point about where the keep the land incrementing code, there are ongoing discussions about where to keep many of the DA-adjacent scripts (for example [g-w #1839](https://github.com/NOAA-EMC/global-workflow/issues/1839)), so I'd imagine that the land incrementing code would fall under similar standards. If you would like to be part of those discussions, let me or other key DA folks (Cory/Daryl/etc) know.

In that case, you should be able to run the new snow DA without my updating the DA code here. For the soil analysis, I do think it's important to start using it i your testing soon (since it's going to change the mean soil moisture, so we should be tuning the model with it included, to avoid what happened last time with the last-minute introduction of the GDAS precip corrections). @GeorgeGayno-NOAA Did we decide to submit my changes as a separate PR, or am I holding up this PR? I'll start on it tomorrow, so I should be able to turn it around pretty quickly.

**Comment by ClaraDraper-NOAA** [comment]
@GeorgeGayno-NOAA The soil analysis on fractional grids looks OK to me, and runs fine once I removed the check preventing it.  

I notice you had a commit dealing with inconsistencies in the various masks. For the soil analysis, the mask is set in calculate_landinc_mask. It will update the STC and SMC if SMC<1.0 and has no snow (since 1 is the value used for SMC over non-soil grids). It also excludes grid cells where the vegetation type indicates glacier. This seems fine to me - do you see any possible issues there?

Note: The new snow analysis will not be OK with fractional grids, but as I said before we're doing the new snow analysis from a different code base. There is already a check preventing calling the new snow analysis with Noah-MP. If you want to be extra careful you could add in an extra check for fractional grids to the new snow analysis.

Note: the current snow analysis will not work with fractional grids, nor with Noah-MP. We should probably add a check to prevent calling it in these cases, since I suspect it will run, but it won't do what is expected.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA The soil analysis on fractional grids looks OK to me, and runs fine once I removed the check preventing it.
> 
> I notice you had a commit dealing with inconsistencies in the various masks. For the soil analysis, the mask is set in calculate_landinc_mask. It will update the STC and SMC if SMC<1.0 and has no snow (since 1 is the value used for SMC over non-soil grids). It also excludes grid cells where the vegetation type indicates glacier. This seems fine to me - do you see any possible issues there?

The logic in calculate_landinc_mask should work for fractional grids as any point with land should have a valid SMC, snow and vegetation type. 
> 
> Note: The new snow analysis will not be OK with fractional grids, but as I said before we're doing the new snow analysis from a different code base. There is already a check preventing calling the new snow analysis with Noah-MP. If you want to be extra careful you could add in an extra check for fractional grids to the new snow analysis.
> 
> Note: the current snow analysis will not work with fractional grids, nor with Noah-MP. We should probably add a check to prevent calling it in these cases, since I suspect it will run, but it won't do what is expected.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA The soil analysis on fractional grids looks OK to me, and runs fine once I removed the check preventing it.
> 
> I notice you had a commit dealing with inconsistencies in the various masks. For the soil analysis, the mask is set in calculate_landinc_mask. It will update the STC and SMC if SMC<1.0 and has no snow (since 1 is the value used for SMC over non-soil grids). It also excludes grid cells where the vegetation type indicates glacier. This seems fine to me - do you see any possible issues there?
> 
> Note: The new snow analysis will not be OK with fractional grids, but as I said before we're doing the new snow analysis from a different code base. There is already a check preventing calling the new snow analysis with Noah-MP. If you want to be extra careful you could add in an extra check for fractional grids to the new snow analysis.
> 
> Note: the current snow analysis will not work with fractional grids, nor with Noah-MP. We should probably add a check to prevent calling it in these cases, since I suspect it will run, but it won't do what is expected.

Added logic at 64e138b

**Pull request merged**